src/MulticallerWithSigner.sol

```diff
@@ -0,0 +1,420 @@
+            // Compute and store `keccak256(abi.encodePacked(message))`.
+            mstore(0x20, keccak256(add(message, 0x20), mload(message)))
+            // Compute and store `keccak256(abi.encodePacked(targets))`.
+            calldatacopy(0x40, targets.offset, data.length)
+            mstore(0x40, keccak256(0x40, data.length))
+            // Compute and store
+            // `keccak256(abi.encodePacked(keccak256(data[0]), .., keccak256(data[n-1])))`.
+            for { let i := returndatasize() } iszero(eq(i, data.length)) { i := add(i, 0x20) } {
+                let o := add(data.offset, calldataload(add(data.offset, i)))
+                let p := add(0x60, i)
+                calldatacopy(p, add(o, 0x20), calldataload(o))
+                mstore(p, keccak256(p, calldataload(o)))
+            }
+            mstore(0x60, keccak256(0x60, data.length))
+            // Compute and store `keccak256(abi.encodePacked(values))`.
+            calldatacopy(0x80, values.offset, data.length)
+            mstore(0x80, keccak256(0x80, data.length))
+            // Store `nonce` and `nonceSalt`.
+            mstore(0xa0, nonce)
+            mstore(0xc0, nonceSalt)
+            mstore(0x40, keccak256(returndatasize(), 0xe0)) // Store the struct hash.
+
+            // `keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)")`.
+            mstore(0x80, 0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f)
+            // `keccak256("MulticallerWithSigner")`.
+            mstore(0xa0, 0x301013e8a31863902646dc218ecd889c37491c2967a8104d5ff1cf42af0f9ea4)
+            // `keccak256("1")`.
+            mstore(0xc0, 0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc6)
+            mstore(0xe0, chainid())
+            mstore(0x100, address())
+
+            // Compute the digest.
+            mstore(returndatasize(), 0x1901) // Store "\x19\x01".
+            mstore(0x20, keccak256(0x80, 0xa0)) // Store the domain separator.
+
+            // Recover the signer from the `digest` and `signature`.
+            mstore(returndatasize(), keccak256(0x1e, 0x42)) // `digest`.
+            mstore(0x20, byte(returndatasize(), calldataload(add(signature.offset, 0x40)))) // `v`.
+            calldatacopy(0x40, signature.offset, 0x40) // Copy `r` and `s`.
+            pop(
+                staticcall(
+                    gas(), // Amount of gas left for the transaction.
+                    and(
+                        // If the signature is exactly 65 bytes in length.
+                        eq(signature.length, 65),
+                        // If `s` in lower half order, such that the signature is not malleable.
+                        lt(
+                            mload(0x60),
+                            0x7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a1
+                        )
+                    ), // Address of `ecrecover`.
```

Comment:
my first reaction was that it is weird to perform validation here, but at the same time it's pretty neat that and'ing all the conditions either results in 0 or 1 (aka the address of ecrecover) 👌